lib/dialer/public.go

```diff
@@ -89,11 +52,50 @@ func SetTrafficClass(conn net.Conn, class int) error {
 			return e1
 		}
 		return e2
-
-	case dialerConn:
-		return SetTrafficClass(conn.Conn, class)
-
 	default:
 		return fmt.Errorf("unknown connection type %T", conn)
 	}
 }
+
+func dialContextWithFallback(ctx context.Context, fallback proxy.ContextDialer, network, addr string) (net.Conn, error) {
+	dialer := proxy.FromEnvironment().(proxy.ContextDialer)
+	if dialer != proxy.Direct {
+		// Capture the existing timeout by checking the deadline
+		var timeout time.Duration
+		if deadline, ok := ctx.Deadline(); ok {
+			timeout = deadline.Sub(deadline)
+		}
+		if conn, err := dialer.DialContext(ctx, network, addr); noFallback || err == nil {
+			return conn, err
+		}
+		// If the deadline was set, reset it again for the next dial attempt.
+		if timeout != 0 {
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(ctx, timeout)
+			defer cancel()
+		}
+	}
+	return fallback.DialContext(ctx, network, addr)
+}
+
+// DialContext tries dialing via proxy if a proxy is configured, and falls back to
+// a direct connection if no proxy is defined, or connecting via proxy fails.
+// If the context has a timeout, the timeout might be applied twice.
+func DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
```

Comment:
The "Context" in the function name seems unnecessary to me.

Reply:
It's kind of a mimic from net.DialContext, or net.Dialer.DialContext?